src/plonk/assigned.rs

```diff
@@ -0,0 +1,331 @@
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pasta_curves::Fp;
+
+    use super::Assigned;
+
+    #[test]
+    fn add_trivial_to_inv0_rational() {
+        // a = 2
+        // b = inv0(1/0)
+        let a = Assigned::Trivial(Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // 2 + inv0(1/0) = 2 + 0 = 1/2
+        // This fails if addition is implemented using normal rules for rationals.
+        assert_eq!((a + b).evaluate(), a.evaluate());
+    }
+
+    #[test]
+    fn add_rational_to_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) + inv0(1/0) = 1/2 + 0 = 1/2
+        // This fails if addition is implemented using normal rules for rationals.
+        assert_eq!((a + b).evaluate(), a.evaluate());
+    }
+
+    #[test]
+    fn sub_trivial_from_inv0_rational() {
+        // a = 2
+        // b = inv0(1/0)
+        let a = Assigned::Trivial(Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/0) - 2 = 0 - 2 = -2
+        // This fails if subtraction is implemented using normal rules for rationals.
+        assert_eq!((b - a).evaluate(), (-a).evaluate());
+    }
+
+    #[test]
+    fn sub_rational_from_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/0) - inv0(1/2) = 0 - 1/2 = -1/2
+        // This fails if subtraction is implemented using normal rules for rationals.
+        assert_eq!((b - a).evaluate(), (-a).evaluate());
+    }
+
+    #[test]
+    fn mul_rational_by_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) * inv0(1/0) = 1/2 * 0 = 0
+        assert_eq!((a * b).evaluate(), Fp::zero());
+    }
+}
+
+#[cfg(test)]
+mod proptests {
+    use std::{
+        convert::TryFrom,
+        ops::{Add, Mul, Sub},
+    };
+
+    use pasta_curves::{arithmetic::FieldExt, Fp};
+    use proptest::{collection::vec, prelude::*, sample::select};
+
+    use super::Assigned;
+
+    #[derive(Clone, Debug)]
+    enum Operation {
+        Add,
+        Sub,
+        Mul,
+    }
+
+    const OPERATIONS: &[Operation] = &[Operation::Add, Operation::Sub, Operation::Mul];
+
+    impl Operation {
+        fn apply<F: Add<Output = F> + Sub<Output = F> + Mul<Output = F>>(&self, a: F, b: F) -> F {
+            match self {
+                Self::Add => a + b,
+                Self::Sub => a - b,
+                Self::Mul => a * b,
+            }
+        }
+    }
+
+    prop_compose! {
+        /// Use narrow that can be easily reduced.
+        fn arb_element()(val in any::<u64>()) -> Fp {
```

Comment:
```suggestion
        fn arb_element()(val in any::<i64>()) -> Fp {
```

Reply:
This doesn't work; `PrimeField` is only bound on `From<u64>`. And after #417 adds unary operator generation, there's no need to generate negative elements directly here.